### PR TITLE
fix(builder): make the test-environment rules boundaries, not conventions

### PR DESCRIPTION
**Pull request by mobeenabdullah**
`packages/builder` and `packages/plugin-page-builder` configured no vitest setup file, so two rules were left to every test file to remember. Both fail **silently and in the reassuring direction**: a file that forgets either one goes on passing.

## What was wrong

**Renders were never torn down between cases.** Nothing registered the testing library's cleanup, so a file without its own `afterEach` accumulates every render into ONE document — and `querySelector` then answers with the FIRST case's element. Later cases assert against a tree belonging to a test that has already finished, and agree with it.

**`act` was not allowed to run.** `React.act` refuses unless `IS_REACT_ACT_ENVIRONMENT` is set, and the refusal is a **warning, not a failure**. A file missing it drives nothing, asserts against the first render, and passes. Exactly one file in the repository set it, and had to set it itself.

## What this does

Both packages load a setup file that acts only where a DOM exists, so it stays inert for the `node`-environment files that are most of each package. That is not a guess about vitest: `packages/plugin-form-builder` already does exactly this, and its config comment states the setup "is inert for the node ones."

The file stays at the package root rather than moving into `src` to gain a tsconfig project. **Measured, not assumed:** it imports `vitest`, and the builder's layering guard counts anything in `src` that is not a test as source, so placing it there fails that guard. Leaving one file unlinted is the better trade than relaxing a guard, and the ignore list now says so.

## Evidence

A setup file's effect is not observable from any ordinary suite going green — which is the shape it exists to prevent. So each package gains a test asserting what the setup file **produces**, from a file that deliberately declares neither rule itself.

Break-verified: removing the `setupFiles` entry fails them, and fails them for the right reasons —

```
FAIL  starts with the previous case's render already torn down
  AssertionError: expected <p data-testid="left-behind"></p> to be null
FAIL  allows React.act to actually run
  AssertionError: expected undefined to be true
```

The first failure names the *actual element left behind by the previous test*, which is the defect itself.

## Blast radius

Baseline before, measured on `origin/main`: builder 71 files / 2075 tests, plugin-page-builder 41 / 484. After: 72 / 2078 and 42 / 487 — the difference is exactly the two new boundary tests. **Nothing regressed and nothing was unmasked.**

Coordinated with the two other lanes live in `packages/builder/**` before touching anything; both confirmed no test of theirs relies on the DOM accumulating between cases.

One honest note: `BlocksField.breakpoints.test.tsx` no longer declares the act flag itself, because the setup now provides it and its comment explaining the local declaration would otherwise be untrue. That file passes with or without the flag, so this is a comment-correctness change, not a bug fix, and I am not claiming otherwise.

### CODE REVIEW

**Comment by mobeenabdullah** [comment]
@codex review

**Comment by mobeenabdullah** [comment]
CI red on `Lint / Typecheck / Test / Build`, and it is **not this change**. Diagnosed rather than re-run, because a green re-run here would prove nothing.

```
FAIL  packages/ui/src/ui-surface.test.ts > ui release tags reach the published types
Error: Hook timed out in 120000ms.
  ❯ src/ui-surface.test.ts:360:3
    360|   beforeAll(() => {
    361|     builtDir = ensureDeclarations();
    362|   }, 120_000);
Test Files  1 failed | 48 passed (49)
     Tests  1129 passed | 6 skipped (1135)
```

Three things make this attributable:

1. **It is a hook timeout, not an assertion.** Every one of the 1129 tests that ran passed. The suite never got to assert anything.
2. **This diff touches zero files in `packages/ui`.** Full file list: two `vitest.config.ts`, two new `vitest.setup.ts`, two new test files, one `eslint-config/base.js` ignore entry, one changeset, and one comment removal in a page-builder test.
3. **It is already recorded**, as `finding:ui-surface-hook-timeout-reds-main`: that hook runs two tsup builds unconditionally inside a 120s budget, so on a runner already executing the rest of the turbo graph it exceeds the wall clock. It reds `main` itself and every PR that inherits it. The recorded remedy is to stop paying a build inside a test hook — build as a task dependency so turbo caches it — rather than raise the number, which only moves the threshold. That is `packages/ui/**` territory and not claimed here.

The one thing I can honestly say this change contributes: it adds two test files, which adds load. Two trivial jsdom files against a 120-second budget is not a plausible cause, but I would rather name it than have it discovered.

Re-running the failed job to get a verdict on the rest of the graph. If it goes green that is not evidence the timeout is fixed — only that the runner was less loaded.

File: .changeset/vitest-setup-boundary.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+---
+"nextly": patch
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply .changeset/vitest-setup-boundary.md]: Confirmed and removed. I checked the rule rather than taking the finding: `AGENTS.md` says "Test-only, CI-only, or docs-only PRs get NO changeset", and I had applied the ONE-changeset-per-PR half without the exemption beside it.

There was a genuine complication, and resolving it is what made the answer clean rather than a judgement call. This PR was not strictly test-only: it also added an ignore entry to `packages/eslint-config/base.js`, which IS a published package, so a consumer of that config would have seen `vitest.setup.*` stop being linted.

That entry existed because the setup file sat at each package root, outside every tsconfig project. Moving it to `scripts/` — which your other finding asked for anyway — puts it somewhere already ignored, so the eslint-config change is gone entirely. `git diff origin/main -- packages/eslint-config/base.js` is now empty.

So the PR touches no published behaviour at all, and `node scripts/release/check-changesets.mjs` agrees: "No changesets added or edited; the lockstep group matches the workspace."

File: scripts/vitest-dom-setup.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * in which suites they guard.
+ */
+import { afterEach } from "vitest";
+
+if ("document" in globalThis) {
+  (
+    globalThis as { IS_REACT_ACT_ENVIRONMENT?: boolean }
+  ).IS_REACT_ACT_ENVIRONMENT = true;
+  const { cleanup } = await import("@testing-library/react");
+  afterEach(cleanup);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply scripts/vitest-dom-setup.ts:57]: Confirmed and fixed, and you are right that this was the worst possible thing to duplicate. Both rules fail silently — a file that has lost either one goes on passing — so a fix applied to one package and not the other produces no signal anywhere. The duplication would not have been discovered by anything failing.

There is now one file, `scripts/vitest-dom-setup.ts`, loaded by both configs by relative path.

I had a reason for the copies and I want to state why it was wrong rather than just drop it. My reasoning was that sharing meant a new module inside `packages/`, which in a 25-package lockstep group is a new published entry for thirty lines no consumer ever loads — and that `plugin-form-builder` already keeps its own. The mistake was treating "inside a package" and "duplicated" as the only options. It is dev tooling, so it belongs outside every package, and `scripts/` is where this repo already puts exactly that.

That placement paid for itself twice over:

- `scripts/**` is already outside every tsconfig project and already ignored by the shared lint config, so the ignore entry I had added to `packages/eslint-config/base.js` is no longer needed — reverted, and that was the only published file in the PR.
- With it gone the change touches no published behaviour, which is what makes your changeset finding cleanly correct rather than a judgement call.

It also keeps the constraint that sent the file to a package root in the first place: it imports `vitest`, so it cannot live under a `src` whose layering guard counts anything that is not a test as source. Measured — I tried `src/__tests__/setup.ts` first and the builder's guard rejected it.

File: packages/plugin-page-builder/src/vitest-setup-boundary.test.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    /*
+     * `React.act` refuses without this flag and the refusal is a warning, so a
+     * suite missing it drives nothing and asserts against the first render.
+     * Read from the global rather than from a rendered effect because the
+     * failure mode is precisely that driving React quietly does nothing.
+     */
+    expect(
+      (globalThis as { IS_REACT_ACT_ENVIRONMENT?: boolean })
+        .IS_REACT_ACT_ENVIRONMENT
+    ).toBe(true);
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/vitest-setup-boundary.test.tsx]: Confirmed and fixed, and this is the one worth dwelling on, because the file it appears in exists to stop exactly this failure and then committed it.

The case asserted that the setup file had set the value the setup file sets. Circular, and green regardless of whether the flag still means anything to React — which is precisely the property the other two cases in that file are written to prevent. I wrote a boundary test for silent-green defects and made one of its three cases silently green.

Your correction is also a fact I had wrong, and I verified it rather than taking it: `act` DOES still execute and flush without the flag. Measured — with the setup removed, the DOM assertion below still passes and only the diagnostic appears. So the flag controls the report, not the behaviour, and the observable thing to assert is the absence of the complaint.

The case now renders a counter, drives it through a bare `React.act`, and asserts both halves: that the update reached the DOM, and that React raised no "not configured to support act" complaint. Break-verified by removing `setupFiles`:

```
FAIL  lets React.act drive an update, with no environment complaint
AssertionError: expected [ Array(1) ] to deeply equal []
```

One array element — React's actual complaint — where the fixed version has none. The DOM assertion passes in both, which is your point demonstrated rather than repeated.

Worth noting what this does NOT now cover, since the original case implied it did: nothing here asserts that a test which forgets the flag produces a WRONG result, because it does not — it produces a correct result and a warning nobody reads. The value of the flag is that the warning becomes visible, and the value of this case is that the warning is absent.

**Comment by mobeenabdullah** [comment]
@codex review

Head is now `75232699b`. All three fixed, and they turned out to be one change rather than three.

Sharing the setup file (finding 2) sent it to `scripts/`, which is already outside every tsconfig project and already ignored by the shared lint config — so the `packages/eslint-config/base.js` entry I had added is gone. That was the only published file in the PR, which makes the changeset finding (1) cleanly correct rather than a judgement call: the change now touches no published behaviour at all, and `check-changesets.mjs` agrees.

**Finding 3 is the one I would flag to a reviewer**, because the file it appears in exists to prevent exactly this and then did it. The case asserted that the setup file had set the value the setup file sets — circular, and green whether or not the flag still means anything to React. A boundary test written against silent-green defects, with one of its three cases silently green.

Your fact-correction was also right and I verified it rather than accepting it: `act` does still execute and flush without the flag; what the flag controls is the diagnostic. So the case now drives a real update through a bare `React.act` and asserts both that it reached the DOM and that React raised no environment complaint. Break-verified — removing `setupFiles` fails it with one complaint present, while the DOM assertion passes in both, which is your point demonstrated.

Gates: builder 2078/2078, plugin-page-builder 487/487, lint clean, and the boundary cases still fail correctly when the setup is removed.

Separately, the earlier CI red on this branch was `packages/ui/src/ui-surface.test.ts` timing out in a `beforeAll` that runs two tsup builds — `finding:ui-surface-hook-timeout-reds-main`, unrelated to this diff, which touches no file in that package.

File: packages/plugin-page-builder/src/vitest-setup-boundary.test.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    /*
+     * This file registers no `afterEach(cleanup)` of its own. Without the setup
+     * file nothing unmounts, the render above is still in this document, and
+     * every `querySelector` in this file answers with THAT element — a case
+     * asserting against a test that has already finished, and agreeing.
+     */
+    expect(document.body.querySelector("[data-testid]")).toBeNull();
+  });
+
+  it("lets React.act drive an update, with no environment complaint", () => {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/plugin-page-builder/src/vitest-setup-boundary.test.tsx:43]: Fixed before this PR merged, and answered late — recording that plainly rather than letting the reply date imply otherwise.

The suites could not be shared, and I measured it both ways round rather than concluding it:

- A module outside every package resolves its imports from the repository root, where `react` is not a dependency. Adding it there would be worse than the duplication it removes: the packages would load their own copy and the shared module another, and two React copies in one renderer means hooks called against the wrong dispatcher.
- Importing such a module from a package's `src` fails `rootDir` — TS6059, "is not under rootDir" — and widening `rootDir` changes what the published build emits.

So the duplication stayed and divergence was made LOUD instead: `vitest-setup-boundary-parity.test.ts` reads both files and fails when they stop matching, ignoring only the `@module` line that names each file's own path. Break-verified with a one-word change to one of them.

That is the same move the suites themselves make, which is why I was comfortable with it as an answer to your finding rather than a dodge: the failure mode you identified was silent, and it was given something that fails.

File: scripts/vitest-dom-setup.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * registered its cleanup. Nothing did. A file without its own `afterEach`
+ * therefore accumulates every render into ONE document, and `querySelector`
+ * answers with the FIRST case's element — so later cases assert against a tree
+ * belonging to a test that has already finished, and agree with it.
+ *
+ * ## `act` is allowed to run
+ *
+ * `React.act` refuses to run unless `IS_REACT_ACT_ENVIRONMENT` is set, and the
+ * refusal is a WARNING rather than a failure. A file missing it drives nothing,
+ * asserts against the first render, and passes. The testing library sets the
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply scripts/vitest-dom-setup.ts]: Confirmed and fixed before merge; answering late, which I should have done at the time.

You were right that it was the canonical place for the wrong belief to live. I had corrected it in the test and left the same claim standing in the setup helper, which is precisely where the next person would read it and recreate the circular assertion.

The docblock now says what was measured: `React.act` runs and flushes whether or not `IS_REACT_ACT_ENVIRONMENT` is set, and what the flag decides is whether React reports the environment as unconfigured. It also names the actual cost, since "refuses to run" made it sound worse and more visible than it is — the cost is not a test that does nothing, it is a `console.error` that a suite full of expected output scrolls straight past and nothing fails on.
